File: sdk/typescript/src/anthropic.ts
import { LatchError } from "./errors.js";

export interface AnthropicAction {
  tool: string;
  arguments?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface AnthropicDecideOptions {
  signal?: AbortSignal;
}

export interface AnthropicDecision {
  readonly decision: string;
  readonly allowed: boolean;
}

export type AnthropicToolHandler = (input: Record<string, unknown>) => unknown | Promise<unknown>;

export interface AnthropicDecisionClient {
  decide(action: AnthropicAction, options?: AnthropicDecideOptions): Promise<AnthropicDecision>;
}

export interface AnthropicToolAdapterOptions {
  maxCalls?: number;
  maxInputBytes?: number;
  maxOutputBytes?: number;
  replayCapacity?: number;
}

export interface AnthropicToolResult {
  type: "tool_result";
  tool_use_id: string;
  content: string;
}

export class AnthropicToolError extends LatchError {
  override readonly name: string = "AnthropicToolError";
}

export class AnthropicToolProtocolError extends AnthropicToolError {
  override readonly name = "AnthropicToolProtocolError";
}

export class UnknownAnthropicToolError extends AnthropicToolError {
  override readonly name = "UnknownAnthropicToolError";
  constructor(readonly toolUseId: string, readonly tool: string) {
    super(`Anthropic tool ${JSON.stringify(tool)} for tool_use ${JSON.stringify(toolUseId)} has no registered handler; batch not executed`);
  }
}

export class AnthropicToolReplayError extends AnthropicToolError {
  override readonly name = "AnthropicToolReplayError";
  constructor(readonly toolUseId: string) {
    super(`Anthropic tool_use ${JSON.stringify(toolUseId)} was duplicated or replayed; batch not executed`);
  }
}

export class AnthropicToolNotAllowedError extends AnthropicToolError {
  override readonly name = "AnthropicToolNotAllowedError";
  constructor(readonly toolUseId: string, readonly tool: string, readonly decision: AnthropicDecision) {
    super(`Latch decision for Anthropic tool ${JSON.stringify(tool)} tool_use ${JSON.stringify(toolUseId)} is ${decision.decision}; batch not executed`);
  }
}

export class AnthropicToolExecutionError extends AnthropicToolError {
  override readonly name = "AnthropicToolExecutionError";
  constructor(readonly toolUseId: string, readonly tool: string, options: { cause: unknown }) {
    super(`execute Anthropic tool ${JSON.stringify(tool)} tool_use ${JSON.stringify(toolUseId)}: ${errorMessage(options.cause)}`, options);
  }
}

interface ToolUse {
  id: string;
  name: string;
  input: Record<string, unknown>;
}

/** Protects tool_use blocks from a completed Anthropic Messages API response without an Anthropic SDK dependency. */
export class AnthropicToolAdapter {
  readonly #client: AnthropicDecisionClient;
  readonly #handlers: ReadonlyMap<string, AnthropicToolHandler>;
  readonly #maxCalls: number;
  readonly #maxInputBytes: number;
  readonly #maxOutputBytes: number;
  readonly #replayCapacity: number;
  readonly #replayed = new Set<string>();
  readonly #replayOrder: string[] = [];

  constructor(
    client: AnthropicDecisionClient,
    handlers: Readonly<Record<string, AnthropicToolHandler>>,
    options: AnthropicToolAdapterOptions = {},
  ) {
    if (client === null || typeof client !== "object" || typeof client.decide !== "function") {
      throw new TypeError("client must provide decide(action)");
    }
    this.#maxCalls = boundedInteger(options.maxCalls ?? 128, 1, 1024, "maxCalls");
    this.#maxInputBytes = boundedInteger(options.maxInputBytes ?? 1 << 20, 1, 16 << 20, "maxInputBytes");
    this.#maxOutputBytes = boundedInteger(options.maxOutputBytes ?? 4 << 20, 1, 64 << 20, "maxOutputBytes");
    this.#replayCapacity = boundedInteger(options.replayCapacity ?? 10_000, 128, 1_000_000, "replayCapacity");
    if (this.#replayCapacity < this.#maxCalls) {
      throw new TypeError("replayCapacity cannot be smaller than maxCalls");
    }
    const copied = new Map<string, AnthropicToolHandler>();
    for (const [name, handler] of Object.entries(handlers)) {
      validateName(name);
      if (typeof handler !== "function") throw new TypeError(`handler ${JSON.stringify(name)} must be a function`);
      copied.set(name, handler);
    }
    this.#client = client;
    this.#handlers = copied;
  }

  async executeMessage(message: unknown, options: AnthropicDecideOptions = {}): Promise<AnthropicToolResult[]> {
    const root = record(message, "message");
    if (root.type !== undefined && root.type !== "message") {
      throw new AnthropicToolProtocolError("payload is not a Messages API message");
    }
    if (root.role !== undefined && root.role !== "assistant") {
      throw new AnthropicToolProtocolError("only assistant messages carry tool_use blocks");
    }
    const calls: ToolUse[] = [];
    for (const blockValue of array(root.content, "message.content")) {
      const block = record(blockValue, "message.content[]");
      if (block.type !== "tool_use") continue;
      calls.push(this.#parseToolUse(block.id, block.name, block.input));
    }
    const outputs = await this.#execute(calls, options);
    return calls.map((call, index) => ({ type: "tool_result", tool_use_id: call.id, content: outputs[index]! }));
  }

  #parseToolUse(idValue: unknown, nameValue: unknown, inputValue: unknown): ToolUse {
    const id = validateToolUseId(idValue);
    const name = validateName(nameValue);
    const input = record(inputValue, `input for tool_use ${JSON.stringify(id)}`);
    let serialized: string;
    try {
      serialized = JSON.stringify(input);
    } catch (error) {
      throw new AnthropicToolProtocolError(`input for tool_use ${JSON.stringify(id)} is invalid: ${errorMessage(error)}`);
    }
    if (serialized === undefined) {
      throw new AnthropicToolProtocolError(`input for tool_use ${JSON.stringify(id)} is not serializable`);
    }
    if (byteLength(serialized) > this.#maxInputBytes) {
      throw new AnthropicToolProtocolError(`input for tool_use ${JSON.stringify(id)} exceeds ${this.#maxInputBytes} bytes`);
    }
    validateJSON(input);
    return { id, name, input };
  }

  async #execute(calls: readonly ToolUse[], options: AnthropicDecideOptions): Promise<string[]> {
    if (calls.length > this.#maxCalls) {
      throw new AnthropicToolProtocolError(`tool_use batch exceeds ${this.#maxCalls} blocks`);
    }
    const seen = new Set<string>();
    for (const call of calls) {
      if (seen.has(call.id)) throw new AnthropicToolReplayError(call.id);
      seen.add(call.id);
      if (!this.#handlers.has(call.name)) throw new UnknownAnthropicToolError(call.id, call.name);
    }
    for (const call of calls) {
      const decision = await this.#client.decide({
        tool: call.name,
        arguments: call.input,
        metadata: {
          protocol: "anthropic-messages",
          surface: "messages",
          tool_use_id: call.id,
          tool_call_kind: "tool_use",
        },
      }, options);
      if (!decision.allowed) throw new AnthropicToolNotAllowedError(call.id, call.name, decision);
    }
    this.#reserve(calls);
    const outputs: string[] = [];
    for (const call of calls) {
      try {
        const value = await this.#handlers.get(call.name)!(call.input);
        outputs.push(encodeOutput(value, this.#maxOutputBytes));
      } catch (error) {
        if (error instanceof AnthropicToolExecutionError) throw error;
        throw new AnthropicToolExecutionError(call.id, call.name, { cause: error });
      }
    }
    return outputs;
  }

  #reserve(calls: readonly ToolUse[]): void {
    for (const call of calls) if (this.#replayed.has(call.id)) throw new AnthropicToolReplayError(call.id);
    for (const call of calls) {
      this.#replayed.add(call.id);
      this.#replayOrder.push(call.id);
    }
    while (this.#replayOrder.length > this.#replayCapacity) {
      this.#replayed.delete(this.#replayOrder.shift()!);
    }
  }
}

function validateToolUseId(value: unknown): string {
  if (typeof value !== "string" || byteLength(value) < 1 || byteLength(value) > 256) {
    throw new AnthropicToolProtocolError("tool_use id must be 1-256 UTF-8 bytes");
  }
  for (const character of value) {
    const code = character.codePointAt(0)!;
    if (code < 0x21 || code > 0x7e) throw new AnthropicToolProtocolError("tool_use id must contain visible ASCII only");
  }
  return value;
}

function validateName(value: unknown): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new AnthropicToolProtocolError("tool name must be 1-256 UTF-8 bytes");
  }
  validateUnicode(value, "tool name");
  if (byteLength(value) > 256) throw new AnthropicToolProtocolError("tool name must be 1-256 UTF-8 bytes");
  return value;
}

function validateJSON(value: unknown): void {
  if (typeof value === "string") {
    validateUnicode(value, "JSON string");
  } else if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new AnthropicToolProtocolError("JSON number is outside the finite range");
    if (Number.isInteger(value) && !Number.isSafeInteger(value)) {
      throw new AnthropicToolProtocolError("JSON integer is outside the interoperable safe range");
    }
  } else if (Array.isArray(value)) {
    for (const item of value) validateJSON(item);
  } else if (value !== null && typeof value === "object") {
    for (const [key, item] of Object.entries(value)) {
      validateUnicode(key, "JSON object key");
      validateJSON(item);
    }
  } else if (typeof value === "bigint" || typeof value === "function" || typeof value === "symbol" || typeof value === "undefined") {
    throw new AnthropicToolProtocolError(`unsupported JSON value of type ${typeof value}`);
  }
}

function encodeOutput(value: unknown, limit: number): string {
  let output: string;
  if (typeof value === "string") output = value;
  else {
    const encoded = JSON.stringify(value);
    if (encoded === undefined) throw new TypeError("tool output is not JSON serializable");
    output = encoded;
  }
  if (byteLength(output) > limit) throw new TypeError(`tool output exceeds ${limit} bytes`);
  return output;
}

function boundedInteger(value: number, minimum: number, maximum: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < minimum || value > maximum) {
    throw new TypeError(`${name} must be an integer between ${minimum} and ${maximum}`);
  }
  return value;
}

function byteLength(value: string): number {
  return new TextEncoder().encode(value).byteLength;
}

function record(value: unknown, name: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new AnthropicToolProtocolError(`${name} must be an object`);
  }
  return value as Record<string, unknown>;
}

function array(value: unknown, name: string): unknown[] {
  if (!Array.isArray(value)) throw new AnthropicToolProtocolError(`${name} must be an array`);
  return value;
}

function errorMessage(value: unknown): string {
  return value instanceof Error ? value.message : String(value);
}

function validateUnicode(value: string, name: string): void {
  for (const character of value) {
    const code = character.codePointAt(0)!;
    if (code === 0xfffd || (character.length === 1 && code >= 0xd800 && code <= 0xdfff)) {
      throw new AnthropicToolProtocolError(`${name} cannot contain replacement or unpaired-surrogate characters`);
    }
  }
}

File: sdk/python/src/latch_sdk/anthropic.py
"""Fail-closed adapter for completed Anthropic Messages API tool_use blocks."""

from collections import OrderedDict
from dataclasses import dataclass
import json
from threading import Lock
from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence

from .errors import LatchError
from .models import Action, Decision
from .openai import _validate_json, _validate_unicode

AnthropicToolHandler = Callable[[Mapping[str, Any]], Any]


class AnthropicToolProtocolError(LatchError):
    """A completed message did not satisfy the protected tool_use contract."""


class UnknownAnthropicToolError(LatchError):
    def __init__(self, tool_use_id: str, tool: str) -> None:
        self.tool_use_id = tool_use_id
        self.tool = tool
        super().__init__(
            f"Anthropic tool {tool!r} for tool_use {tool_use_id!r} has no registered handler; batch not executed"
        )


class AnthropicToolReplayError(LatchError):
    def __init__(self, tool_use_id: str) -> None:
        self.tool_use_id = tool_use_id
        super().__init__(f"Anthropic tool_use {tool_use_id!r} was duplicated or replayed; batch not executed")


class AnthropicToolNotAllowedError(LatchError):
    def __init__(self, tool_use_id: str, tool: str, decision: Decision) -> None:
        self.tool_use_id = tool_use_id
        self.tool = tool
        self.decision = decision
        super().__init__(
            f"Latch decision for Anthropic tool {tool!r} tool_use {tool_use_id!r} is {decision.decision}; batch not executed"
        )


class AnthropicToolExecutionError(LatchError):
    def __init__(self, tool_use_id: str, tool: str, cause: Exception) -> None:
        self.tool_use_id = tool_use_id
        self.tool = tool
        self.cause = cause
        super().__init__(f"execute Anthropic tool {tool!r} tool_use {tool_use_id!r}: {cause}")


@dataclass(frozen=True)
class _ToolUse:
    tool_use_id: str
    name: str
    input: Mapping[str, Any]


class AnthropicToolAdapter:
    """Protect Anthropic Messages API tool_use blocks with Latch.

    Every tool_use block is parsed, resolved to a handler, and allowed before
    any handler starts. Block IDs are then atomically consumed, so redelivery
    of the same assistant message cannot repeat a side effect.
    """

    def __init__(
        self,
        client: Any,
        handlers: Mapping[str, AnthropicToolHandler],
        *,
        max_calls: int = 128,
        max_input_bytes: int = 1 << 20,
        max_output_bytes: int = 4 << 20,
        replay_capacity: int = 10_000,
    ) -> None:
        if client is None or not callable(getattr(client, "decide", None)):
            raise TypeError("client must provide decide(action)")
        if isinstance(max_calls, bool) or not isinstance(max_calls, int) or not 1 <= max_calls <= 1024:
            raise ValueError("max_calls must be between 1 and 1024")
        if not 1 <= max_input_bytes <= 16 << 20:
            raise ValueError("max_input_bytes must be between 1 and 16777216")
        if not 1 <= max_output_bytes <= 64 << 20:
            raise ValueError("max_output_bytes must be between 1 and 67108864")
        if not 128 <= replay_capacity <= 1_000_000 or replay_capacity < max_calls:
            raise ValueError("replay_capacity must be between 128 and 1000000 and at least max_calls")
        copied: Dict[str, AnthropicToolHandler] = {}
        for name, handler in handlers.items():
            _validate_name(name)
            if not callable(handler):
                raise TypeError(f"handler {name!r} must be callable")
            copied[name] = handler
        self._client = client
        self._handlers = copied
        self._max_calls = max_calls
        self._max_input_bytes = max_input_bytes
        self._max_output_bytes = max_output_bytes
        self._replay_capacity = replay_capacity
        self._replayed: "OrderedDict[str, None]" = OrderedDict()
        self._replay_lock = Lock()

    def execute_message(self, message: Any) -> List[Mapping[str, str]]:
        """Execute tool_use blocks from one completed assistant message.

        Returns tool_result content blocks to append to a new user message.
        """
        message_type = _field(message, "type", required=False)
        if message_type is not None and message_type != "message":
            raise AnthropicToolProtocolError("payload is not a Messages API message")
        role = _field(message, "role", required=False)
        if role is not None and role != "assistant":
            raise AnthropicToolProtocolError("only assistant messages carry tool_use blocks")
        calls: List[_ToolUse] = []
        for block in _array(_field(message, "content"), "message.content"):
            if _field(block, "type") != "tool_use":
                continue
            calls.append(
                self._parse_tool_use(
                    _field(block, "id"), _field(block, "name"), _field(block, "input")
                )
            )
        results = self._execute(calls)
        return [
            {"type": "tool_result", "tool_use_id": call.tool_use_id, "content": results[index]}
            for index, call in enumerate(calls)
        ]

    def _parse_tool_use(self, tool_use_id: Any, name: Any, raw_input: Any) -> _ToolUse:
        _validate_tool_use_id(tool_use_id)
        _validate_name(name)
        if not isinstance(raw_input, Mapping):
            raise AnthropicToolProtocolError(f"input for tool_use {tool_use_id!r} must be an object")
        try:
            serialized = json.dumps(raw_input, ensure_ascii=False, allow_nan=False)
        except (TypeError, ValueError) as error:
            raise AnthropicToolProtocolError(f"input for tool_use {tool_use_id!r} is invalid: {error}") from error
        if len(serialized.encode("utf-8")) > self._max_input_bytes:
            raise AnthropicToolProtocolError(
                f"input for tool_use {tool_use_id!r} exceeds {self._max_input_bytes} bytes"
            )
        input_object = dict(raw_input)
        try:
            _validate_json(input_object)
        except ValueError as error:
            raise AnthropicToolProtocolError(f"input for tool_use {tool_use_id!r} is invalid: {error}") from error
        return _ToolUse(tool_use_id, name, input_object)

    def _execute(self, calls: Sequence[_ToolUse]) -> List[str]:
        if len(calls) > self._max_calls:
            raise AnthropicToolProtocolError(f"tool_use batch exceeds {self._max_calls} blocks")
        seen = set()
        for call in calls:
            if call.tool_use_id in seen:
                raise AnthropicToolReplayError(call.tool_use_id)
            seen.add(call.tool_use_id)
            if call.name not in self._handlers:
                raise UnknownAnthropicToolError(call.tool_use_id, call.name)
        for call in calls:
            decision = self._client.decide(
                Action(
                    tool=call.name,
                    arguments=call.input,
                    metadata={
                        "protocol": "anthropic-messages",
                        "surface": "messages",
                        "tool_use_id": call.tool_use_id,
                        "tool_call_kind": "tool_use",
                    },
                )
            )
            if not decision.allowed:
                raise AnthropicToolNotAllowedError(call.tool_use_id, call.name, decision)
        self._reserve(calls)
        outputs: List[str] = []
        for call in calls:
            try:
                value = self._handlers[call.name](call.input)
                outputs.append(self._encode_output(value))
            except AnthropicToolExecutionError:
                raise
            except Exception as error:
                raise AnthropicToolExecutionError(call.tool_use_id, call.name, error) from error
        return outputs

    def _reserve(self, calls: Sequence[_ToolUse]) -> None:
        with self._replay_lock:
            for call in calls:
                if call.tool_use_id in self._replayed:
                    raise AnthropicToolReplayError(call.tool_use_id)
            for call in calls:
                self._replayed[call.tool_use_id] = None
            while len(self._replayed) > self._replay_capacity:
                self._replayed.popitem(last=False)

    def _encode_output(self, value: Any) -> str:
        if isinstance(value, str):
            encoded = value
        else:
            encoded = json.dumps(value, separators=(",", ":"), ensure_ascii=False, allow_nan=False)
        if len(encoded.encode("utf-8")) > self._max_output_bytes:
            raise ValueError(f"tool output exceeds {self._max_output_bytes} bytes")
        return encoded


def _field(value: Any, name: str, *, required: bool = True) -> Any:
    if isinstance(value, Mapping):
        if name in value:
            return value[name]
    elif hasattr(value, name):
        return getattr(value, name)
    if required:
        raise AnthropicToolProtocolError(f"required field {name!r} is missing")
    return None


def _array(value: Any, name: str) -> Sequence[Any]:
    if isinstance(value, (str, bytes, bytearray)) or not isinstance(value, Sequence):
        raise AnthropicToolProtocolError(f"{name} must be an array")
    return value


def _validate_tool_use_id(value: Any) -> None:
    if not isinstance(value, str):
        raise AnthropicToolProtocolError("tool_use id must be 1-256 UTF-8 bytes")
    if any(ord(character) < 0x21 or ord(character) > 0x7E for character in value):
        raise AnthropicToolProtocolError("tool_use id must contain visible ASCII only")
    if not 1 <= len(value) <= 256:
        raise AnthropicToolProtocolError("tool_use id must be 1-256 UTF-8 bytes")


def _validate_name(value: Any) -> None:
    if not isinstance(value, str) or not value.strip():
        raise AnthropicToolProtocolError("tool name must be 1-256 UTF-8 bytes")
    try:
        _validate_unicode(value, "tool name")
    except ValueError as error:
        raise AnthropicToolProtocolError("tool name must contain valid interoperable Unicode") from error
    if len(value.encode("utf-8")) > 256:
        raise AnthropicToolProtocolError("tool name must be 1-256 UTF-8 bytes")
